### PR TITLE
Support helm3 in configure-rbac

### PR DESCRIPTION
**Pull request by poelzi**
helm3 has different calling syntax then helm2 for rendering templates
only. Detect if running under helm2 or 3 and use the proper syntax

**How to categorize this PR?**

/area dev-productivity
/kind enhancement

**What this PR does / why we need it**:

support for helm3

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator

```

### CODE REVIEW

**Comment by gardener-robot-ci-1** [comment]
Thank you @poelzi for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

**Comment by timebertt** [comment]
Thanks @poelzi. I get the motivation of the change, as I also ran into this, so I opened https://github.com/gardener/gardener/issues/3961.
I'm a bit reluctant to introduce a switch here and there for helm 2 vs 3, deal with this in all places and have inconsistent handling all over the place. I would rather like to see all usages of `helm` in our hack scripts adapted all at once, so we can completely switch to helm 3 everywhere.
So far, I haven't found the time to do it. Would you like to tackle https://github.com/gardener/gardener/issues/3961?

**Comment by poelzi** [comment]
@timebertt I updated the install scripts to install version 3.5.4 and tested what is located in the hack folder. I removed the version check

File: hack/generate-controller-registration.sh
Comment (timebertt) [inline hack/generate-controller-registration.sh:50]: I think we can also remove line 69 now:
https://github.com/gardener/gardener/blob/14b4c3730c492fbd86e8841cbe40f9ac82b8e4b4/hack/generate-controller-registration.sh#L69

**Review by timebertt — COMMENTED** [review]
@poelzi thanks for updating the PR.
I now cross-checked the change with the aws extension and it seems to work mostly fine (`generate-controller-registration.sh` works and `ControllerRegistration` is installable/upgradable in a gardener installation).

However there is one issue when running `make check` (the `check-charts` part):
```
> Check Helm charts
Checking for chart symlink errors
Checking whether all charts can be rendered
walk.go:74: found symbolic link in path: /Users/d067603/go/src/github.com/gardener/gardener-extension-provider-aws/charts/gardener-extension-admission-aws/charts/application/values.yaml resolves to /Users/d067603/go/src/github.com/gardener/gardener-extension-provider-aws/charts/gardener-extension-admission-aws/values.yaml
walk.go:74: found symbolic link in path: /Users/d067603/go/src/github.com/gardener/gardener-extension-provider-aws/charts/gardener-extension-admission-aws/charts/runtime/values.yaml resolves to /Users/d067603/go/src/github.com/gardener/gardener-extension-provider-aws/charts/gardener-extension-admission-aws/values.yaml
Error: YAML parse error on aws-infra/templates/main.tf: error converting YAML to JSON: yaml: line 427: mapping values are not allowed in this context

Use --debug flag to render out invalid YAML
```
It complains when checking the `terraform` chart https://github.com/gardener/gardener-extension-provider-aws/tree/master/charts/internal/aws-infra, which we use to template the terraform config, as the resources there are not valid yaml.
I see two options for dealing with this:
- either exclude the terraform config chart from being checked in `check-charts`
- or completely remove the chart (it's a bit hacky anyway) and replace it with some plain go template (should be sufficient for the use case)

(similar things would need to be done in the other provider extensions using terraform as well)

@gardener/gardener-maintainers opinions on this one?

Side-note: I also checked the [helm v3 migration guide](https://helm.sh/docs/topics/v2_v3_migration) and it seems like there are no breaking changings wrt the templating engine itself (which is mainly the part of helm we actually use), so I assume, proceeding like this with the upgrade should be fine.

**Comment by rfranzke** [comment]
@timebertt I don't get what the issue with the TF chart of AWS is - I don't even find this line:

> Error: YAML parse error on aws-infra/templates/main.tf: error converting YAML to JSON: yaml: line 427: mapping values are not allowed in this context

Can you explain?

**Comment by timebertt** [comment]
helm 3 complains, if the rendered document is not valid YAML (e.g. when using `helm template`). Then you get the above error.

**Comment by rfranzke** [comment]
Aha, interesting. Then I guess we should go for using native Go templates and `embed` instead of the chart (this also fits to what we do with #2754 in g/g). However, that's quite some effort as it needs to be done for most provider extensions...

**Comment by timebertt** [comment]
> However, that's quite some effort as it needs to be done for most provider extensions...

True. That's why I'm asking for more opinions :)
Should we still go ahead with the helm upgrade now and add the adaption part in the extensions to the backlog (there's some time left until we'll vendor the next gardener release)?

**Comment by rfranzke** [comment]
I would rather vote for reversing this, i.e., first adapt the extensions, and when we are happy then we can go ahead and update Helm to v3. This way there is no "pressure" to adapt.

**Comment by timebertt** [comment]
Ok, @poelzi would you be fine with holding your PR until then?

**Comment by poelzi** [comment]
The first version of my patch was fine with just checking which version of helm was running and using the appropriate switch depending on the helm version. I'm not using the install requirements script because I'm running nixos and will not clutter my system with a script downloading executables somewhere.

**Comment by rfranzke** [comment]
Yeah, I agree, the demand is out of scope of @poelzi's PR. Hence, to proceed, I'd propose to go back to the initial version of this PR and have the switch in the `configure-rbac` script so that it works with both Helm 2/3. In parallel (but independently), we should try to prepare the extensions for the Helm 3 update.

**Comment by rfranzke** [comment]
FYI: I've opened a PR for the AWS extension which replaces the "infra" Helm chart with native Golang templates: https://github.com/gardener/gardener-extension-provider-aws/pull/332

**Comment by timebertt** [comment]
> FYI: I've opened a PR for the AWS extension which replaces the "infra" Helm chart with native Golang templates: [gardener/gardener-extension-provider-aws#332](https://github.com/gardener/gardener-extension-provider-aws/pull/332)

Now that we have this as an example, shouldn't we just go ahead with this PR as is? It shouldn't be too much effort to adapt the other extensions similarly, right?
(we also still have ~2 weeks left before revendoring this g/g version)

**Comment by rfranzke** [comment]
Depends on whether the extensions maintainers have time and capacity to adapt... Generally, I think it produces avoidable "pressure", but let's ask them: @kon-angelo @dkistner @prashanth26 @minchaow Do you think you can do something similar like https://github.com/gardener/gardener-extension-provider-aws/pull/332 for the other provider extensions within the next week?

Let's
/hold
until this is clarified.

**Comment by prashanth26** [comment]
> Depends on whether the extensions maintainers have time and capacity to adapt... Generally, I think it produces avoidable "pressure", but let's ask them: @kon-angelo @dkistner @prashanth26 @minchaow Do you think you can do something similar like [gardener/gardener-extension-provider-aws#332](https://github.com/gardener/gardener-extension-provider-aws/pull/332) for the other provider extensions within the next week?
> 
> Let's
> /hold
> until this is clarified.

Hi Guys,

I have some priority feature releases on other components (like a serialized reattachment of PVs and more) to make this week which will probably eat up most of my time. I know this change is probably not too difficult to make, however, I would like to avoid committing to this change by EO this week. For me, it would be great if I could get a couple of weeks from today to incorporate and test this change. However, if others are okay with the above deadline and if this is something needed on priority, I can try (no promises) to make the change later this week.

Thanks & Regards,
Prashanth

**Comment by dkistner** [comment]
Same for me.
I have some priority dev task to continue in the next ~1-2 weeks and I will most likely not have time to look into it this week.
Same is probably true for @kon-angelo, but he wanted to explore a little bit what would be required to do, but he has also some major development activities in front.

**Comment by kon-angelo** [comment]
My current task involves quite a bit of refactoring for the terraform files in Azure and I wanted to see what will be missing when we swap to the go templates. I ended up making the PR for Azure, but as the other colleagues said, there other high priority tasks in the works and I would hope that we get a bit more time.

**Comment by rfranzke** [comment]
@poelzi @timebertt WDYT about rolling back the PR to the initial version (switch between v2/v3) for now? After the colleagues have adapted the extensions we can delete the v2 parts.

**Review by timebertt — APPROVED** [review]
Let's get this in :) 
/lgtm

**Review by rfranzke — APPROVED** [review]
/lgtm
/unhold

Thanks for your help, colleagues!

**Pull request merged**
